File: typescript/src/detector/high/origins-verified-cross-origin-communications/origins-verified-cross-origin-communications.ts
// {fact rule=origins-verified-cross-origin-communications@v1.0 defects=1}
function originsVerifiedCrossOriginCommunicationsNoncompliant() {
  var iframe = document?.querySelector(".testiframe") as HTMLIFrameElement;
  // Noncompliant: the wildcard keyword `*` is used.
  iframe?.contentWindow?.postMessage("secret_value", "*");
}
// {/fact}

// {fact rule=origins-verified-cross-origin-communications@v1.0 defects=0}
function originsVerifiedCrossOriginCommunicationsCompliant() {
  var iframe = document?.querySelector(".testiframe") as HTMLIFrameElement;
  // Compliant: using secure origin.
  iframe?.contentWindow?.postMessage(
    "secret_value",
    "https://secure.example.com",
  );
}

// {/fact}

File: csharp/src/detectors/origins-verified-cross-origin-communications/OriginsVerifiedCrossOriginCommunications.cs
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

namespace CrossOrigin
{
    class OriginsVerifiedCrossOriginCommunications
    {
        // {fact rule=origins-verified-cross-origin-communications@v1.0 defects=1}
        public void ConfigureNoncompliant(IServiceCollection service)
        {
            // Noncompliant: The HSTS HTTP response security header is missing.
            service.AddControllers();
        }
        // {/fact}
        
        // {fact rule=origins-verified-cross-origin-communications@v1.0 defects=0}
        public void ConfigureCompliant(IServiceCollection service)
        {
            service.AddControllers();
            // Compliant: `AddHsts` HTTP response security header is used.
            service.AddHsts(options =>
            {
                options.Preload = true;
                options.IncludeSubDomains = true;
                options.MaxAge = TimeSpan.FromDays(365);
            });
        }
        // {/fact}
    }
}

